test/extended/images/mysql_replica.go

```diff
@@ -0,0 +1,311 @@
+package images
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// getPods looks up pods with "oc get" applying a label selector and returns
+// the names of matching pods
+func getPods(oc *exutil.CLI, selector string) ([]string, error) {
+	outPods, err := oc.Run("get").Args("pods",
+		"-l", selector,
+		"-o", "template",
+		"-t", "{{range .items}}{{.metadata.name}},{{end}}").Output()
+	if err != nil {
+		return nil, err
+	}
+	return strings.Split(strings.Trim(outPods, ","), ","), nil
+}
+
+// pollForPods will call getPods in a loop until the required number of pods
+// is returned
+func pollForPods(oc *exutil.CLI, selector string, count int, timeout time.Duration) ([]string, error) {
+	var pods []string
+	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
+		p, e := getPods(oc, selector)
+		if e != nil {
+			return true, e
+		}
+		if len(p) < count {
+			return false, nil
+		}
+		pods = p
+		return true, nil
+	})
+	return pods, err
+}
+
+type mysql struct {
+	oc        *exutil.CLI
+	ExecPod   string
+	TargetPod string
+	Endpoint  string
+
+	// ENV vars
+	User         string
+	UserPassword string
+	Database     string
+	RootPassword string
+}
+
+func NewMysql(oc *exutil.CLI, execPod, targetPod, service string) (*mysql, error) {
+	mysql := &mysql{
+		oc:        oc,
+		ExecPod:   execPod,
+		TargetPod: targetPod,
+		Endpoint:  service + "." + oc.Namespace() + ".svc.cluster.local",
+	}
+
+	out, err := oc.Run("env").Args("pod", targetPod, "--list").Output()
+	if err != nil {
+		return nil, err
+	}
+
+	lines := strings.Split(out, "\n")
+	for _, line := range lines {
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+		s := strings.SplitN(line, "=", 2)
+		switch s[0] {
+		case "MYSQL_USER":
+			mysql.User = s[1]
+		case "MYSQL_PASSWORD":
+			mysql.UserPassword = s[1]
+		case "MYSQL_DATABASE":
+			mysql.Database = s[1]
+		case "MYSQL_ROOT_PASSWORD":
+			mysql.RootPassword = s[1]
+		case "MYSQL_MASTER_USER", "MYSQL_MASTER_PASSWORD":
+			// Ignore
+		default:
+			return nil, errors.New("Unknown environment variable named " + s[0])
+		}
+	}
+	return mysql, nil
+}
+
+// IsUp pings the MySQL server
+func (m *mysql) IsUp() (bool, error) {
+	out, err := m.oc.Run("exec").ExecBashArgs(m.TargetPod, "mysqladmin -h "+"127.0.0.1"+" -u"+"root"+" ping").Output()
+	if err != nil {
+		switch err.(type) {
+		case *exutil.ExitError:
+			return false, nil
+		default:
+			return false, err
+		}
+	}
+	return out == "mysqld is alive", nil
+}
+
+// WaitUntilUp continuously waits for the server to respond to pings, up until timeout.
```

Comment:
Does this method waits only for ping ? couldnt it depend on some other mechanism ?